services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -113,38 +167,287 @@ class HakemusServiceITest : DatabaseTest() {
+                )
+        }
+
+        @Test
+        fun `throws exception when area is not inside hanke area`() {
+            val hanke = hankeFactory.builder(USERID).withHankealue().saveEntity()
+            val application = createHakemus(hanke)
+
+            val exception =
+                assertThrows<ApplicationGeometryNotInsideHankeException> {
+                    hakemusService.updateHakemus(
+                        application.id!!,
+                        HakemusUpdateRequestFactory
+                            .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(
+                                application
+                            )
+                            .copy(areas = listOf(notInHankeArea)),
+                        USERID
+                    )
+                }
+
+            Assertions.assertEquals(
+                """Application geometry doesn't match any hankealue when updating application id=${application.id},
+                | hankeId = ${hanke.id},
+                | application geometry = ${notInHankeArea.geometry.toJsonString()}"""
+                    .trimMargin()
+                    .replace("\n", ""),
+                exception.message
+            )
+        }
+
+        @Test
+        fun `saves updated data and creates an audit log`() {
+            val hanke = hankeFactory.builder(USERID).withHankealue().saveEntity()
+            val application =
+                createHakemusWithHakija(hanke) {
+                    applicationData =
+                        (applicationData as CableReportApplicationData).copy(
+                            workDescription = "Old work description"
+                        )
+                }
+            val yhteystieto = hakemusyhteystietoRepository.findAll().first()
+            val kayttaja =
+                hankekayttajaRepository
+                    .findByHankeIdAndSahkopostiIn(hanke.id, listOf(KAYTTAJA_INPUT_HAKIJA.email))
+                    .single()
+            val newKayttaja = hankeKayttajaFactory.saveUser(hanke.id)
+            val originalAuditLogSize = auditLogRepository.findByType(ObjectType.APPLICATION).size
+
+            val request =
+                HakemusUpdateRequestFactory
+                    .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(application)
+                    .copy(
+                        // change work description
+                        workDescription = "New work description",
+                        // add a new contact
+                        customerWithContacts =
+                            HakemusUpdateRequestFactory.createCustomerWithContactsRequest(
+                                CustomerType.COMPANY,
+                                yhteystieto.id,
+                                kayttaja.id,
+                                newKayttaja.id
+                            )
+                    )
```

Comment:
This is one place where generation can be made simpler by using extension methods. The copy can be hidden inside the extension method, making the result a lot cleaner.
```suggestion
                    .withCustomer(CustomerType.COMPANY, yhteystieto.id, kayttaja.id, newKayttaja.id)
                    .withDescription("New work description")
```

The extension functions are simple and can be added to `HakemusUpdateRequestFactory`:
```kotlin
    fun JohtoselvityshakemusUpdateRequest.withCustomer(
        customerType: CustomerType = CustomerType.COMPANY,
        yhteystietoId: UUID? = UUID.randomUUID(),
        vararg hankekayttajaIds: UUID
    ) =
        copy(
            customerWithContacts =
                createCustomerWithContactsRequest(customerType, yhteystietoId, *hankekayttajaIds)
        )

    fun JohtoselvityshakemusUpdateRequest.withDescription(description: String) =
        copy(workDescription = description)
```

Reply:
Thank you!